lib/std/comptime_hash_map.zig

```diff
@@ -0,0 +1,140 @@
+const std = @import("std.zig");
+const hash_map = std.hash_map;
+const testing = std.testing;
+const math = std.math;
+
+/// A comptime hashmap constructed with automatically selected hash and eql functions.
+pub fn AutoComptimeHashMap(comptime K: type, comptime V: type, comptime values: var) type {
+    return ComptimeHashMap(K, V, hash_map.getAutoHashFn(K), hash_map.getAutoEqlFn(K), values);
+}
+
+/// Builtin hashmap for strings as keys.
+pub fn ComptimeStringHashMap(comptime V: type, comptime values: var) type {
+    return ComptimeHashMap([]const u8, V, hash_map.hashString, hash_map.eqlString, values);
+}
+
+/// A hashmap which is constructed at compile time from constant values.
+/// Intended to be used as a faster lookup table.
+pub fn ComptimeHashMap(comptime K: type, comptime V: type, comptime hash: fn (key: K) u32, comptime eql: fn (a: K, b: K) bool, comptime values: var) type {
+    std.debug.assert(values.len != 0);
+    @setEvalBranchQuota(1000 * values.len);
+
+    const Entry = struct {
+        distance_from_start_index: usize = 0,
+        key: K = undefined,
+        val: V = undefined,
+        used: bool = false,
+    };
+
+    // ensure that the hash map will be at most 60% full
+    const size = math.ceilPowerOfTwo(usize, values.len * 5 / 3) catch unreachable;
+    var slots = [1]Entry{.{}} ** size;
+
+    var max_distance_from_start_index = 0;
+
+    slot_loop: for (values) |kv| {
+        var key: K = kv[0];
+        var value: V = kv[1];
```

Comment:
From messing around with making the implementation in #5442 re-usable (https://github.com/ziglang/zig/pull/5442#issuecomment-634321596), I found that using `[0]` and `[1]` here restricts `kv` to being an anonymous list literal (or an array, presumably, but that would require K and V to be the same type). This limits the use cases since (AFAIK) it's not possible to generate arrays/slices of anonymous list literals at compile time, which would be needed for using `ComptimeHashMap` in something like a `std.meta.stringToEnum` implementation.

Using `kv.@"0"` and `kv.@"1"` might be better (since that will work with both anonymous list literals and struct types with fields specifically named `@"0"` and `@"1"`), or maybe make it conditional on `std.meta.trait.isIndexable` if arrays are intended to be supported as well?